### PR TITLE
JBRes-1867 Action

### PR DESCRIPTION
**Pull request by SBOne-Kenobi**
Here are the main changes:
- Added an action to execute project minimization
- Updated versions of the IntelliJ IDEA and dependencies
- Revised coroutines usage
- Some functional updates to be able to execute minimization correctly

P.S. It's planned to merge this PR after the PR #7 about error filtering

### CODE REVIEW

File: project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt [hunk 1]
```diff
@@ -1,33 +1,110 @@
...
+            val level = smartReadAction(context.project) {
+                val roots = ProjectRootManager.getInstance(context.project).contentSourceRoots
+                    .takeIf { it.isNotEmpty() } ?: arrayOf(context.projectDir)
+
+                context.progressReporter?.let {
+                    reporter = ProgressReporter(it, context.projectDir.toNioPath(), roots)
+                }
+
+                roots.map { ProjectFileDDItem.create(context, it) }
+            }
```
Comment (khbminus) [inline project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt:35]: Does this also solve [JBRES-1791](https://youtrack.jetbrains.com/issue/JBRes-1791/Gradle-Hierarchy-Producer)?

File: project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt [hunk 2]
```diff
@@ -1,33 +1,110 @@
...
+                        entry.next += 1
+                    } else {
+                        stack.removeLast()
+                        levelMaxDepths[entry.file.toNioPath().relativeTo(root)] =
+                            children.maxOf { levelMaxDepths[it.toNioPath().relativeTo(root)]!! }
+                    }
+                } else {
+                    stack.removeLast()
+                    levelMaxDepths[entry.file.toNioPath().relativeTo(root)] = entry.level
+                }
```
Comment (khbminus) [inline project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt]: I kinda do not understand what is going here. What is `next`?

File: project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/services/ProjectCloningService.kt [hunk 3]
```diff
@@ -33,24 +33,33 @@ class ProjectCloningService(private val rootProject: Project) {
...
         return ProjectUtil.openOrImportAsync(clonedProjectPath.toNioPath())
     }
 
-    @Suppress("UnstableApiUsage")
-    private suspend fun createNewProjectDirectory(): VirtualFile = writeAction {
-        getSnapshotLocation().findOrCreateDirectory(UUID.randomUUID().toString())
+    private val importantFiles = setOf("modules.xml", "misc.xml", "libraries")
+    private fun ignore(file: VirtualFile, root: VirtualFile): Boolean {
+        val path = file.toNioPath().relativeTo(root.toNioPath())
+        if (path.startsWith(".idea")) {
```
Comment (khbminus) [inline project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/services/ProjectCloningService.kt]: I suggest using `Project.projectPath` here

**Review by khbminus — APPROVED** [review]
LGTM :)

File: project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt (hunk 1 above)
Comment (SBOne-Kenobi) [reply project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt:35]: Partially yes

File: project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt (hunk 2 above)
Comment (SBOne-Kenobi) [reply project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/hierarchy/FileTreeHierarchicalDDGenerator.kt]: That's the index of the next child to iterate. It's a recursion simulation. 

I will rename it and add some comments

**Pull request merged**